llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -1031,6 +1031,27 @@ void State::addInfoForInductions(BasicBlock &BB) {
   WorkList.push_back(FactOrCheck::getConditionFact(
       DTN, CmpInst::ICMP_SLT, PN, B,
       ConditionTy(CmpInst::ICMP_SLE, StartValue, B)));
+
+  assert(!StepOffset.isNegative() && "induction must be increasing");
+  // Try to add condition from header to the unique exit block, if there is one.
+  // When exiting either with EQ or NE, we know that the induction value must be
+  // u<= B, as a different exit may exit earlier.
+  ConditionTy Precond;
+  if (!MonotonicallyIncreasingUnsigned)
+    Precond = {CmpInst::ICMP_ULE, StartValue, B};
```

Comment:
I'm confused, why do we only need this pre-condition if `!MonotonicallyIncreasingUnsigned`? Let's say we have a monotonic nuw addrec, a header eq exit where the start value is greater that is not taken and some other exit that is taken. Wouldn't we miscompile that?

Reply:
Doesn't `MonotonicallyIncreasingUnsigned` imply that `StartValue u<= ExitValue == B`?

Reply:
I don't think so. It just implies that PN is monotonically increasing from StartValue, but it doesn't make a statement about how that relates to B.

If I take your first test case from https://alive2.llvm.org/ce/z/UYsbAQ, the current patch still folds it to true.

Reply:
Thanks, updated to always add the precond, originally also thought that `MonotonicallyIncreasingUnsigned` was implying `StartValue <= B`. Test added in 3d11b3d750b735a149ab47b6ad8b8f98929c36a5